tests/server_send.cpp

```diff
@@ -0,0 +1,349 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ */
+
+#include "test_base.hpp"
+#include "../library/elliptics.h"
+#include <algorithm>
+
+#define BOOST_TEST_NO_MAIN
+#include <boost/test/included/unit_test.hpp>
+
+#include <boost/program_options.hpp>
+
+using namespace ioremap::elliptics;
+using namespace boost::unit_test;
+
+static std::shared_ptr<tests::nodes_data> ssend_servers;
+
+static std::vector<int> ssend_src_groups = {1};
+static std::vector<int> ssend_dst_groups = {2, 3};
+static size_t ssend_backends = 8;
+
+static std::string print_groups(const std::vector<int> &groups) {
+	std::ostringstream ss;
+	for (size_t pos = 0; pos < groups.size(); ++pos) {
+		ss << groups[pos];
+		if (pos != groups.size() - 1)
+			ss << ":";
+	}
+
+	return ss.str();
+}
+
+static tests::server_config ssend_server_config(int group)
+{
+	// Minimize number of threads
+	tests::server_config server = tests::server_config::default_value();
+	server.options
+		("io_thread_num", 4)
+		("nonblocking_io_thread_num", 4)
+		("net_thread_num", 1)
+		("caches_number", 1)
+	;
+
+	server.backends[0]("enable", true)("group", group);
+	server.backends.resize(ssend_backends, server.backends.front());
+
+	return server;
+}
+
+static void ssend_configure(const std::string &path)
+{
+	std::vector<tests::server_config> servers;
+	for (const auto &g : ssend_src_groups) {
+		tests::server_config server = ssend_server_config(g);
+		servers.push_back(server);
+	}
+	for (const auto &g : ssend_dst_groups) {
+		tests::server_config server = ssend_server_config(g);
+		servers.push_back(server);
+	}
+
+	tests::start_nodes_config cfg(results_reporter::get_stream(), std::move(servers), path);
+	cfg.fork = true;
+
+	ssend_servers = tests::start_nodes(cfg);
+}
+
+static void ssend_test_insert_many_keys(session &s, int num, const std::string &id_prefix, const std::string &data_prefix)
+{
+	s.set_trace_id(rand());
+	for (int i = 0; i < num; ++i) {
+		std::string id = id_prefix + lexical_cast(i);
+		std::string data = data_prefix + lexical_cast(i);
+
+		ELLIPTICS_REQUIRE(res, s.write_data(id, data, 0));
+	}
+}
+
+static void ssend_test_read_many_keys(session &s, int num, const std::string &id_prefix, const std::string &data_prefix)
+{
+	s.set_trace_id(rand());
+	for (int i = 0; i < num; ++i) {
+		std::string id = id_prefix + lexical_cast(i);
+		std::string data = data_prefix + lexical_cast(i);
+
+		ELLIPTICS_COMPARE_REQUIRE(res, s.read_data(id, 0, 0), data);
+	}
+}
+
+static void ssend_test_read_many_keys_error(session &s, int num, const std::string &id_prefix, int error)
+{
+	s.set_trace_id(rand());
+	for (int i = 0; i < num; ++i) {
+		std::string id = id_prefix + lexical_cast(i);
+
+		ELLIPTICS_REQUIRE_ERROR(res, s.read_data(id, 0, 0), error);
+	}
+}
+
+static std::vector<dnet_raw_id> ssend_ids(session &s)
```

Comment:
Ok, I thought that `if (std::find(groups.begin(), groups.end(), entry.group_id) != groups.end()) {` implies multiple groups with all unique backen_id.